okio/src/main/java/okio/Pipe.java

```diff
@@ -56,43 +60,128 @@ public final Sink sink() {
+   * with {@code sink}. This pipe's source is closed and attempts to read it will throw an
+   * {@link IllegalStateException}.
+   *
+   * This method must not be called while concurrently accessing this pipe's source. It is safe,
+   * however, to call this while concurrently writing this pipe's sink.
+   */
+  void fold(Sink sink) throws IOException {
+    while (true) {
+      Buffer sinkBuffer = null;
+      synchronized (buffer) {
+        if (foldedSink != null) throw new IllegalStateException("sink already folded");
+
+        if (buffer.exhausted()) {
+          sourceClosed = true;
+          foldedSink = sink;
+          timeoutFolder = new TimeoutFolder(sink.timeout());
+          return;
+        }
+
+        sinkBuffer = new Buffer();
+        sinkBuffer.write(buffer, buffer.size);
+        buffer.notifyAll(); // Notify the sink that it can resume writing.
+      }
+
+      boolean success = false;
+      try {
+        sink.write(sinkBuffer, sinkBuffer.size);
+        success = true;
+      } finally {
+        if (!success) {
+          synchronized (buffer) {
+            sourceClosed = true;
+            buffer.notifyAll(); // Notify the sink that it can resume writing.
+          }
+        }
+      }
+    }
+  }
+
   final class PipeSink implements Sink {
     final Timeout timeout = new Timeout();
 
     @Override public void write(Buffer source, long byteCount) throws IOException {
+      Sink delegate = null;
       synchronized (buffer) {
         if (sinkClosed) throw new IllegalStateException("closed");
 
-        while (byteCount > 0) {
-          if (sourceClosed) throw new IOException("source is closed");
-
-          long bufferSpaceAvailable = maxBufferSize - buffer.size();
-          if (bufferSpaceAvailable == 0) {
-            timeout.waitUntilNotified(buffer); // Wait until the source drains the buffer.
-            continue;
+        while (byteCount > 0 && delegate == null) {
+          if (foldedSink != null) {
+            delegate = foldedSink;
+          } else {
+            if (sourceClosed) throw new IOException("source is closed");
+
+            long bufferSpaceAvailable = maxBufferSize - buffer.size();
+            if (bufferSpaceAvailable == 0) {
+              timeout.waitUntilNotified(buffer); // Wait until the source drains the buffer.
+              continue;
+            }
+
+            long bytesToWrite = Math.min(bufferSpaceAvailable, byteCount);
+            buffer.write(source, bytesToWrite);
+            byteCount -= bytesToWrite;
+            buffer.notifyAll(); // Notify the source that it can resume reading.
           }
+        }
+      }
 
-          long bytesToWrite = Math.min(bufferSpaceAvailable, byteCount);
-          buffer.write(source, bytesToWrite);
-          byteCount -= bytesToWrite;
-          buffer.notifyAll(); // Notify the source that it can resume reading.
+      if (delegate != null) {
+        timeoutFolder.push(this.timeout());
```

Comment:
Regard to `timeoutFolder`, should I do something similar to what we do for getting `folderSink` out of the synchronized block?

Reply:
yes please. It’s the easiest to reason about